UNIX/SOURCE/xgrabsc-2.2/xgrab.man
.\"========================================================================
.\"
.\" Name - xgrab.man
.\"
.\" ccs version: 1.6
.\"
.\" ccsid:	@(#)xgrab.man	1.6 - 8/12/92 12:39:23
.\" from: 	ccs/s.xgrab.man
.\" date: 	8/12/92 14:15:41
.\"
.\" Copyright (C) 1990-92 Bruce Schuchardt
.\" See the end of this document for full copyright information.
.\"
.\" Description:  Man page for xgrab
.\"
.\"========================================================================
.\"
.TH XGRAB 1X
.\"
.SH NAME
xgrab \- interactive front for xgrabsc, an X-Windows image grabber
.\"
.SH SYNOPSIS
.in +8n
.ti -8n
\fIxgrab\fR
.in -8n
.\"
.SH DESCRIPTION
\fIxgrab\fR lets you grab arbitrary rectangular images from an
X server and writes them to files or commands (such as lpr) in
a variety of formats.
.PP
\fIxgrab\fR is a front for the xgrabsc program.
Read the man page for \fIxgrabsc\fR for a description of the options
presented by xgrab.
.PP
After selecting options from the various categories presented, press
the \fIOK\fR button to have xgrab run xgrabsc to let you grab an image
from the screen.  After you press \fIOK\fP, xgrab's window will disappear
and xgrabsc will gain control until the grabbing process is finished.
Afterwards, the xgrab window will reappear.
.sp 2
.SH OPTIONS
Xgrab responds to the standard application options, such as \fI-display\fP.
See the man page for \fIX\fP for a complete list.
.sp
You can also override the default xgrab settings in your .Xdefaults.  See
the examples section below for instructions.
.sp 2
.SH RESOURCES
The XGrab resource file, XGrab.ad, contains a complete specification of the
resources of all the widgets used in the xgrab window.  Global resources,
such as default font and color, are at the bottom of the file.
.sp 2
.SH EXAMPLES
The \fITo Command\fR output option may be used to pipe xgrabsc output
to programs.  The most common commands are \fIlpr\fR for Postscript
output, and \fIxwud\fR for X-Window Dump output.  Programs that do not
accept piped input should not be used in \fITo Command\fR.
.sp
You can also get fancy and pipe the output through more than one command,
such as \fItee screen.dmp | xwud\fP, to store the grabbed image and get
a preview window.
.sp 2
Default settings for xgrab can be made in your .Xdefaults file.  Toggle
buttons can be set/unset through their \fI.state\fP attribute, and text
field strings can be set through their \fI*string\fP attribute.  For
example, to set the default paper size for postscript output, put these
lines in your .Xdefaults (and use xrdb to load them into the server!):
.sp
XGrab*.pageWidthText*string:   8.5
.br
XGrab*.pageHeightText*string: 11.0
.sp
To set the default output type to \fIXWD\fP, put these lines in
your .Xdefaults:
.sp
XGrab*.ps.state:  0
.br
XGrab*.xwd.state: 1
.sp 2
.SH LIMITATIONS
See the limitations listed in the xgrabsc man page.
.sp 2
.SH AUTHOR
.nf
     Bruce Schuchardt
    Servio Corporation
      bruce@slc.com
.fi
.sp 2
.SH ACKNOWLEGEMENTS
.PP
George Ferguson wrote the ad2c program.
.sp 2
.SH COPYRIGHT
Copyright (c) 1991-92 Bruce Schuchardt
.PP
\fIXgrab\fR is copywritten material with a very loose copyright
allowing unlimited modification and distribution if the copyright
notices are left intact.  Various portions are copywritten by various
people, but all use a modification of the MIT copyright notice.
Please check the source for complete copyright information.  The
intent is to keep the source free, not to stifle its distribution, so
please write to me if you have any questions.
.PP
THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE
USE OR PERFORMANCE OF THIS SOFTWARE.
.s 2
.SH SEE ALSO
X(1X), xgrabsc(1X), xhost(1), xwd(1X), xwud(1X), xloadimage(1X), xpm(1X),
xpr(1X), ad2c(1)


UNIX/SOURCE/xgrabsc-2.2/xgrab.c
/*========================================================================
 *
 * Name - xgrab.c
 *
 * Version:	1.11
 *
 * ccsid:	@(#)xgrab.c	1.11 - 8/12/92 14:14:56
 * from: 	ccs/s.xgrab.c
 * date: 	8/12/92 14:15:41
 *
 * Copyright (c) 1990-92 Bruce Schuchardt.
 * Read the file cpyright.h for full copyright information.
 *
 *
 * Description:
 *
 * xgrab.c - interactive front for xgrabsc
 *
 *========================================================================
 */
#include "cpyright.h"
#include "patchlevel.h"
#include "config.h"

#include <stdio.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Shell.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/Label.h>
#include <X11/Xaw/Command.h>
#include <X11/Xaw/AsciiText.h>
#include <X11/Xaw/Toggle.h>

static char *fallback_resources[] = {
#include "xgrab_ad.h"
  NULL
  };


static Display *hDisplay;

typedef enum {
  HITWIN = 1,
  STRETCH,
  KEY,
  ROOT,

  XWD,
  PS,
  BITMAP,
  BITMAP2,
  BITMAP3,
  PUZZLE,
  SIMPLE,

  DITHER,
  HALFTONE,
  MAPBW,

  REVERSE,

  TOFILE,
  TOPRN,

  PS_NORMAL,
  PS_EPS,
  PS_EPSI,
  PS_EPSIONLY,
  PS_LANDSCAPE
  
  } radioType;


Widget selBtns, formatBtns, psColr, encBtns, psComp, sltext, psltext,
       psLim,
       psBin, borderBtn, horizMargin, vertMargin, pgHeight, pgWidth,
       cnvBtns, revBtn, brtext, outBtns, filenm, prncmd, dispBtn, hostnm;
Widget hShell;

XtAppContext appContext;


static void doGrab() {
  char tmp[257];
  char cmdargs[200];
  char piper[200];
  char cmd[500];
  char *txt;
  int  s, d, len, num;
  int  sltime, psltime;
  int  bright;
  Boolean state;

  cmdargs[0] = '\0';
  piper[0] = '\0';

  XtVaGetValues(sltext, XtNstring, &txt, NULL);
  sltime = atoi(txt);
  if (sltime < 0) sltime = 0;
  sprintf(tmp, "-sleep %d ", sltime);
  strcat(cmdargs, tmp);

  XtVaGetValues(psltext, XtNstring, &txt, NULL);
  psltime = atoi(txt);
  if (psltime < 0) psltime = 0;
  sprintf(tmp, "-post %d ", psltime);
  strcat(cmdargs, tmp);

  XtVaGetValues(brtext, XtNstring, &txt, NULL);
  bright = atoi(txt);
  sprintf(tmp, "-brighten %d ", bright);
  strcat(cmdargs, tmp);

  XtVaGetValues(borderBtn, XtNstate, &state, NULL);
  if (!state)
    strcat(cmdargs, "-");
  strcat(cmdargs, "-bdrs ");

  XtVaGetValues(revBtn, XtNstate, &state, NULL);
  if (!state)
    strcat(cmdargs, "-");
  strcat(cmdargs, "-reverse ");

  switch ((int)XawToggleGetCurrent(selBtns)) {
    case HITWIN:
      strcat(cmdargs, "-click ");
      break;
    case KEY:
      strcat(cmdargs, "-key ");
      break;
    case ROOT:
      strcat(cmdargs, "-root ");
      break;
    case STRETCH:
    default:
      strcat(cmdargs, "-stretch ");
      break;
  }
  switch ((int)XawToggleGetCurrent(formatBtns)) {
    case XWD:
      strcat(cmdargs, "-xwd ");
      break;
    case BITMAP:
      strcat(cmdargs, "-bm ");
      break;
    case BITMAP2:
      strcat(cmdargs, "-bm2 ");
      break;
    case BITMAP3:
      strcat(cmdargs, "-bm3 ");
      break;
    case PUZZLE:
      strcat(cmdargs, "-puzzle ");
      break;
    case SIMPLE:
      strcat(cmdargs, "-simple ");
      break;
    case PS:
    default:
      XtVaGetValues(psColr, XtNstate, &state, NULL);
      if (!state)
        strcat(cmdargs, "-ps ");
      else
        strcat(cmdargs, "-cps ");

      switch ((int)XawToggleGetCurrent(encBtns)) {
        case PS_EPS:
          strcat(cmdargs, "-eps ");
	  break;
	case PS_EPSI:
	  strcat(cmdargs, "-preview ");
	  break;
	case PS_EPSIONLY:
	  strcat(cmdargs, "-previewonly ");
	  break;
	case PS_LANDSCAPE:
	  strcat(cmdargs, "-landscape ");
	  break;
	default:
	  break;
      }

      XtVaGetValues(psComp, XtNstate, &state, NULL);
      if (!state)
        strcat(cmdargs, "-");
      strcat(cmdargs, "-compress ");
      XtVaGetValues(psBin, XtNstate, &state, NULL);
      if (!state)
        strcat(cmdargs, "-");
      strcat(cmdargs, "-bin ");

      XtVaGetValues(psLim, XtNstate, &state, NULL);
      if (!state)
        strcat(cmdargs, "-");
      strcat(cmdargs, "-limit ");

      strcat(cmdargs, "-page ");
      XtVaGetValues(pgWidth, XtNstring, &txt, NULL);
      if (strlen(txt))    strcat(cmdargs, txt);
      else                strcat(cmdargs, "8.5");
      strcat(cmdargs, "x");
      XtVaGetValues(pgHeight, XtNstring, &txt, NULL);
      if (strlen(txt))    strcat(cmdargs, txt);
      else                strcat(cmdargs, "11.0");
      strcat(cmdargs, "-");
      XtVaGetValues(horizMargin, XtNstring, &txt, NULL);
      if (strlen(txt))    strcat(cmdargs, txt);
      else                strcat(cmdargs, "0.5");
      strcat(cmdargs, "-");
      XtVaGetValues(horizMargin, XtNstring, &txt, NULL);
      if (strlen(txt))    strcat(cmdargs, txt);
      else                strcat(cmdargs, "0.5");
      strcat(cmdargs, " ");

      break;
  }
  switch ((int)XawToggleGetCurrent(cnvBtns)) {
    case DITHER:
      strcat(cmdargs, "-dither ");
      break;
    case HALFTONE:
      strcat(cmdargs, "-halftone ");
      break;
    case MAPBW:
      strcat(cmdargs, "-bw ");
      break;
    default:
      break;
  }
  switch ((int)XawToggleGetCurrent(outBtns)) {
    case TOPRN:
      XtVaGetValues(prncmd, XtNstring, &txt, NULL);
      if (strlen(txt)) {
        strcat(piper, " | ");
        strcat(piper, txt);
      }
      break;
    case TOFILE:
    default:
      strcat(cmdargs, "-o ");
      XtVaGetValues(filenm, XtNstring, &txt, NULL);
      if (strlen(txt)) {
        strcat(cmdargs, txt);
#ifdef BUMP_FILENAMES
	len = strlen(txt);
	for (s=0; s<len && (txt[s] != '.'); s++)
	{}
	strcpy(tmp, txt);
	tmp[s] = '\0';
	d = s;
	while (d > 0  &&  ('0' <= txt[d-1]  &&  txt[d-1] <= '9'))
	  d--;
	if ('0' <= txt[d]  && txt[d] <= '9')
	  num = atoi(&tmp[d]);
	else
	  num = 0;
	num++;
	sprintf(&tmp[d], "%d", num);
	strcat(tmp, &txt[s]);
	XtVaSetValues(filenm, XtNstring, tmp, NULL);
#endif
      }
      else {
        strcat(cmdargs, "screen.dmp");
	XtVaSetValues(filenm, XtNstring, "screen.dmp", NULL);
      }
      strcat(cmdargs, " ");
      break;
  }

  XtVaGetValues(dispBtn, XtNstate, &state, NULL);
  if (state) {
    XtVaGetValues(hostnm, XtNstring, &txt, NULL);
    strcat(cmdargs, "-d ");
    if (strlen(txt)) {
      strcat(cmdargs, txt);
      strcat(cmdargs, " ");
    }
    else
      strcat(cmdargs, ":0 ");
  }

  sprintf(cmd, "xgrabsc %s%s\n", cmdargs, piper);
  /* fputs(cmd, stderr); */
  /* try to make the window iconic */
  XtUnmapWidget(hShell);
  XSync(hDisplay, False);
  system(cmd);
  XtMapWidget(hShell);
}




static void doDismiss() {
  if (hShell) {
    XtDestroyWidget(hShell);
    hShell = NULL;
    exit(0);
  }
}




static void createWindow() {
  Widget dialog, button, box1, box2, box3, box4, box5, box6;

  dialog = XtVaCreateManagedWidget("dialog", formWidgetClass, hShell,
           NULL);
           XtVaCreateManagedWidget("title",  labelWidgetClass, dialog,
           NULL);


  /* selection options */
  /* input options */
  box1 = XtVaCreateManagedWidget("box1", formWidgetClass, dialog, NULL);

           XtVaCreateManagedWidget("inputLbl", labelWidgetClass, box1,
	   NULL);

  dispBtn= XtVaCreateManagedWidget("host", toggleWidgetClass, box1,
	   NULL);

  hostnm = XtVaCreateManagedWidget("hostText", asciiTextWidgetClass, box1,
           XtNeditType, XawtextEdit, NULL);

  if (getenv("DISPLAY"))
    XtVaSetValues(hostnm, XtNstring, getenv("DISPLAY"), NULL);

  selBtns =
           XtVaCreateManagedWidget("click", toggleWidgetClass, box1,
           XtNradioData, HITWIN,
	   NULL);

           XtVaCreateManagedWidget("stretch", toggleWidgetClass, box1,
           XtNradioData, STRETCH,
	   XtNradioGroup, selBtns,
	   NULL);

           XtVaCreateManagedWidget("key", toggleWidgetClass, box1,
           XtNradioData, KEY,
	   XtNradioGroup, selBtns,
	   NULL);

           XtVaCreateManagedWidget("root", toggleWidgetClass, box1,
           XtNradioData, ROOT,
	   XtNradioGroup, selBtns,
	   NULL);

           XtVaCreateManagedWidget("sleeplbl", labelWidgetClass, box1,
           NULL);

  sltext = XtVaCreateManagedWidget("sleeptime", asciiTextWidgetClass, box1,
           XtNeditType, XawtextEdit, NULL);

           XtVaCreateManagedWidget("psleeplbl", labelWidgetClass, box1,
           NULL);

  psltext = XtVaCreateManagedWidget("psleeptime", asciiTextWidgetClass, box1,
           XtNeditType, XawtextEdit, NULL);

  box2 = XtVaCreateManagedWidget("box2", formWidgetClass, dialog, NULL);

  /* output options */
           XtVaCreateManagedWidget("outputFormat", labelWidgetClass, box2,
           NULL);

  formatBtns =
           XtVaCreateManagedWidget("xwd", toggleWidgetClass, box2,
           XtNradioData, XWD,
	   NULL);

           XtVaCreateManagedWidget("ps", toggleWidgetClass, box2,
           XtNradioData, PS,
	   XtNradioGroup, formatBtns,
	   NULL);

           XtVaCreateManagedWidget("xpm", toggleWidgetClass, box2,
           XtNradioData, BITMAP,
	   XtNradioGroup, formatBtns,
	   NULL);

           XtVaCreateManagedWidget("xpm2", toggleWidgetClass, box2,
           XtNradioData, BITMAP2,
	   XtNradioGroup, formatBtns,
	   NULL);

           XtVaCreateManagedWidget("xpm3", toggleWidgetClass, box2,
           XtNradioData, BITMAP3,
	   XtNradioGroup, formatBtns,
	   NULL);

           XtVaCreateManagedWidget("puzzle", toggleWidgetClass, box2,
           XtNradioData, PUZZLE,
	   XtNradioGroup, formatBtns,
	   NULL);



  /* postscript options */
  box3 = XtVaCreateManagedWidget("box3", formWidgetClass, dialog, NULL);

           XtVaCreateManagedWidget("psOptions", labelWidgetClass, box3,
           NULL);

  psComp = XtVaCreateManagedWidget("compress", toggleWidgetClass, box3,
	   NULL);

  psColr = XtVaCreateManagedWidget("color", toggleWidgetClass, box3,
	   NULL);

           XtVaCreateManagedWidget("pageWidth", labelWidgetClass, box3,
           NULL);

  pgWidth = XtVaCreateManagedWidget("pageWidthText",asciiTextWidgetClass, box3,
           XtNeditType, XawtextEdit, NULL);

           XtVaCreateManagedWidget("pageHeight", labelWidgetClass, box3,
           NULL);

  pgHeight = XtVaCreateManagedWidget("pageHeightText",asciiTextWidgetClass, box3,
           XtNeditType, XawtextEdit, NULL);

           XtVaCreateManagedWidget("horizMargin", labelWidgetClass, box3,
           NULL);

  horizMargin = XtVaCreateManagedWidget("horizMarginText",asciiTextWidgetClass, box3,
           XtNeditType, XawtextEdit, NULL);

           XtVaCreateManagedWidget("vertMargin", labelWidgetClass, box3,
           NULL);

  vertMargin = XtVaCreateManagedWidget("vertMarginText",asciiTextWidgetClass, box3,
           XtNeditType, XawtextEdit, NULL);


  encBtns= XtVaCreateManagedWidget("landscape", toggleWidgetClass, box3,
           XtNradioData, PS_LANDSCAPE,
	   NULL);

           XtVaCreateManagedWidget("encap", toggleWidgetClass, box3,
           XtNradioData, PS_EPS,
	   XtNradioGroup, encBtns,
	   NULL);

           XtVaCreateManagedWidget("onlyEpsi", toggleWidgetClass, box3,
           XtNradioData, PS_EPSIONLY,
	   XtNradioGroup, encBtns,
	   NULL);

           XtVaCreateManagedWidget("epsi", toggleWidgetClass, box3,
           XtNradioData, PS_EPSI,
	   XtNradioGroup, encBtns,
	   NULL);

  psBin = XtVaCreateManagedWidget("binary", toggleWidgetClass, box3,
	   NULL);

  psLim = XtVaCreateManagedWidget("limit", toggleWidgetClass, box3,
	   NULL);


  /* image processing options */
  box4 = XtVaCreateManagedWidget("box4", formWidgetClass, dialog, NULL);

           XtVaCreateManagedWidget("prOptions", labelWidgetClass, box4,
           NULL);
  cnvBtns= XtVaCreateManagedWidget("dither", toggleWidgetClass, box4,
           XtNradioData, DITHER,
	   NULL);
           XtVaCreateManagedWidget("halftone", toggleWidgetClass, box4,
           XtNradioData, HALFTONE,
	   XtNradioGroup, cnvBtns,
	   NULL);
           XtVaCreateManagedWidget("mapbw", toggleWidgetClass, box4,
           XtNradioData, MAPBW,
	   XtNradioGroup, cnvBtns,
	   NULL);

           XtVaCreateManagedWidget("brightnessLbl", labelWidgetClass, box4,
	   NULL);
  brtext = XtVaCreateManagedWidget("brightnessText", asciiTextWidgetClass, box4,
           XtNeditType, XawtextEdit, NULL);

  revBtn = XtVaCreateManagedWidget("reverse", toggleWidgetClass, box4, NULL);

  borderBtn = XtVaCreateManagedWidget("borders", toggleWidgetClass, box4, NULL);

  box5 = XtVaCreateManagedWidget("box5", formWidgetClass, dialog, NULL);

           XtVaCreateManagedWidget("outputLbl", labelWidgetClass, box5,
	   NULL);

  outBtns= XtVaCreateManagedWidget("file", toggleWidgetClass, box5,
           XtNradioData, TOFILE,
	   NULL);

  filenm = XtVaCreateManagedWidget("fileText", asciiTextWidgetClass, box5,
           XtNeditType, XawtextEdit, NULL);

           XtVaCreateManagedWidget("printer", toggleWidgetClass, box5,
           XtNradioData, TOPRN,
	   XtNradioGroup, outBtns,
	   NULL);

  prncmd = XtVaCreateManagedWidget("printerText", asciiTextWidgetClass, box5,
           XtNeditType, XawtextEdit, NULL);


  box6 = XtVaCreateManagedWidget("box6", formWidgetClass, dialog, NULL);

  XtVaCreateManagedWidget("splat", labelWidgetClass, box6, NULL);
  
  button = XtVaCreateManagedWidget("OK", commandWidgetClass, box6, NULL);
  XtAddCallback(button, XtNcallback, doGrab, NULL);

  button = XtVaCreateManagedWidget("Dismiss", commandWidgetClass, box6, NULL);
  XtAddCallback(button, XtNcallback, doDismiss, NULL);

}



int main(argc, argv, envp)
  int argc;
  char *argv[];
  char *envp;
{
  puts("X-Windows Screen Grabber");
  puts(Copyright);
  puts("");


  hShell = XtAppInitialize(&appContext, "XGrab", NULL, 0, &argc, argv,
     fallback_resources, NULL, 0);

  hDisplay = XtDisplay(hShell);
  createWindow();

  XtRealizeWidget(hShell);
  XtAppMainLoop(appContext);
}
